fix(settings): Unlock Settings mutex before emitting setting changed signals

anthonybilinski opened the pull request:
When a setting is set it locks the Settings mutex and emits a signal for that
setting. If a slot is connected to that signal and lives on the same thread as
Settings, it is executed immediately with the Settings mutex still locked. That
slot can then lock a series of other mutexes. During this time another thread
can lock a mutex and then try to read settings, which will cause a deadlock due
to the opposite order of multiple mutex locking. By making all Settings
connectiong queued, the Settings setter will return and unlock the Settings
mutex, ensuring that getters won't ever be blocked indefinitely.

Fix #6218

- [x] Commits follow our [git commit guidelines](https://github.com/qTox/qTox/blob/master/CONTRIBUTING.md#git-commit-guidelines)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/6219)
<!-- Reviewable:end -->

sudden6 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6219#-:-MFgpba7AoGtwai1-5rt:b-jbsf8z)* status: :shipit: complete! 1 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->

anthonybilinski commented:
Chatting with @sudden6 and @sphaerophoria, a cleaner solution would be to unlock the mutex before emitting setting changes signals inside of Settings. This would make it so that any new user of Settings' signals would need to know to use queued connection type.

I'll update the PR for that.

anthonybilinski committed:
fix(settings): Unlock Settings mutex before emitting setting changed signals

When a setting is set it locks the Settings mutex and emits a signal for that
setting. If a slot is connected to that signal and lives on the same thread as
Settings, it is executed immediately with the Settings mutex still locked. That
slot can then lock a series of other mutexes. During this time another thread
can lock a mutex and then try to read settings, which will cause a deadlock due
to the opposite order of multiple mutex locking. By always emitting signals
after unlocking the Settings mutex, we avoid holding the Settings mutex while
executing slots.

Fix #6218

anthonybilinski committed:
fix(settings): Emit correct signal when spellchecking is updated

anthonybilinski reviewed (commented):
Totally reworked the PR to unlock the mutex before signals are emitted rather than making all connections queued. Requesting re-review <img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6219#-:-MFjeYej5TzXurx0JcWS:b-5886pg)* status: :shipit: complete! 1 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->

sudden6 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6219)* status: :shipit: complete! 1 of 1 LGTMs obtained

___
*[src/persistence/settings.h, line 575 at r2](https://reviewable.io/reviews/qtox/qtox/6219#-MFk5x_6Cs_PMwrvD4vJ:-MFk5x_72GRjpdXHiOqN:b-jbab8a) ([raw file](https://github.com/qtox/qtox/blob/ebc931b1f79478878e3b9d51d8be9b1b2be6e96c/src/persistence/settings.h#L575)):*
> ```objc
>     ICoreSettings::ProxyType fixInvalidProxyType(ICoreSettings::ProxyType proxyType);
> 
>     template <class ValType>
> ```

Why in the header and not in the ´cpp` file?
___
*[src/persistence/settings.cpp, line 919 at r2](https://reviewable.io/reviews/qtox/qtox/6219#-MFk6OZ880I3WUHsWkkm:-MFk6OZ880I3WUHsWkkn:b-qcq5bz) ([raw file](https://github.com/qtox/qtox/blob/cbead167422f405f5a64b5c56c8f978f6e60f1c9/src/persistence/settings.cpp#L919)):*
> ```cpp
> #ifdef QTOX_PLATFORM_EXT
>     QMutexLocker locker{&bigLock};
> ```

why is the mutex removed without any other changes?
___
*[src/persistence/settings.cpp, line 1272 at r2](https://reviewable.io/reviews/qtox/qtox/6219#-MFk6yeQ253Qe202Inq0:-MFk6yeRCoHkkerDZuwq:b-eq4c2x) ([raw file](https://github.com/qtox/qtox/blob/ebc931b1f79478878e3b9d51d8be9b1b2be6e96c/src/persistence/settings.cpp#L1272)):*
> ```cpp
> void Settings::setAutoAwayTime(int newValue)
> {
>     if (newValue < 0)
> ```

while at it, add braces


<!-- Sent from Reviewable.io -->

sudden6 reviewed (commented):
<img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6219#-:-MFkG5ca95_e2ysljgNL:b8b79qu)* status: 0 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->

anthonybilinski reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6219)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/settings.h, line 575 at r2](https://reviewable.io/reviews/qtox/qtox/6219#-MFk5x_6Cs_PMwrvD4vJ:-MFl4xqaDMUz7g2feGfJ:bpqc9da) ([raw file](https://github.com/qtox/qtox/blob/ebc931b1f79478878e3b9d51d8be9b1b2be6e96c/src/persistence/settings.h#L575)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Why in the header and not in the ´cpp` file?
</blockquote></details>

The definition is in the header because of https://isocpp.org/wiki/faq/templates#templates-defn-vs-decl.

The declaration isn't it an anonymous namespace is the `cpp` file because it depends on `bigLock` from `Settings`.
___
*[src/persistence/settings.cpp, line 919 at r2](https://reviewable.io/reviews/qtox/qtox/6219#-MFk6OZ880I3WUHsWkkm:-MFl3xgM3WNCeCZFMXtT:bncuv2u) ([raw file](https://github.com/qtox/qtox/blob/cbead167422f405f5a64b5c56c8f978f6e60f1c9/src/persistence/settings.cpp#L919)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

why is the mutex removed without any other changes?
</blockquote></details>

This whole function touches no `Settings` state.
___
*[src/persistence/settings.cpp, line 1272 at r2](https://reviewable.io/reviews/qtox/qtox/6219#-MFk6yeQ253Qe202Inq0:-MFl4xEZ41AQtEhTVLlN:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/ebc931b1f79478878e3b9d51d8be9b1b2be6e96c/src/persistence/settings.cpp#L1272)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

while at it, add braces
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sudden6 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 1 of 10 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6219#-:-MFlhnOUB0_a4d6nvXAj:b-jbsf8z)* status: :shipit: complete! 1 of 1 LGTMs obtained

___
*[src/persistence/settings.h, line 575 at r2](https://reviewable.io/reviews/qtox/qtox/6219#-MFk5x_6Cs_PMwrvD4vJ:-MFlhXMY0LKVp3VqukWd:b2yq3vv) ([raw file](https://github.com/qtox/qtox/blob/ebc931b1f79478878e3b9d51d8be9b1b2be6e96c/src/persistence/settings.h#L575)):*
<details><summary><i>Previously, anthonybilinski (Anthony Bilinski) wrote…</i></summary><blockquote>

The definition is in the header because of https://isocpp.org/wiki/faq/templates#templates-defn-vs-decl.

The declaration isn't it an anonymous namespace is the `cpp` file because it depends on `bigLock` from `Settings`.
</blockquote></details>

done, overlooked that it's inside the class definition


<!-- Sent from Reviewable.io -->

anthonybilinski reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6219)* status: :shipit: complete! 1 of 1 LGTMs obtained

___
*[src/persistence/settings.h, line 575 at r2](https://reviewable.io/reviews/qtox/qtox/6219#-MFk5x_6Cs_PMwrvD4vJ:-MFmnf1g5T-CmZmdCMxS:bf8uc4a) ([raw file](https://github.com/qtox/qtox/blob/ebc931b1f79478878e3b9d51d8be9b1b2be6e96c/src/persistence/settings.h#L575)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

done, overlooked that it's inside the class definition
</blockquote></details>

oops actually since this is private, sticking it in the cpp still allows all callers to see its definition at call time. This would fail to link to any outside user, but that's not a concern since it's private. Moved implementation to cpp.


<!-- Sent from Reviewable.io -->
